server/src/main/java/io/crate/execution/dml/ObjectIndexer.java

```diff
@@ -214,15 +239,48 @@ private void addNewColumns(Map<String, Object> value,
             );
             innerIndexers.put(innerName, valueIndexer);
             innerTypes.put(innerName, type);
-            xContentBuilder.field(innerName);
-            valueIndexer.indexValue(
+            valueIndexer.collectSchemaUpdates(
                 innerValue,
-                xContentBuilder,
-                addField,
                 onDynamicColumn,
-                synthetics,
-                checks
+                synthetics
             );
         }
     }
+
+    /**
+     * All unknown columns are already handled in the first phase, thus phase 2 is aware of **almost** all unknwn columns.
+     * However, columns added into IGNORED objects and are not handled in the first phase and thus needs to be handked even on the second phase.
+     * Also, empty arrays and arrays filled only by nulls doesn't result in a new column, thus also needs separate handling.
+     */
```

Comment:
```suggestion
    /**
     * Writes keys and values for which there are no columns after {@link #collectSchemaUpdates(Map, Consumer, Map) to the xContentBuilder.
     *
     * There are no columns for:
     * <ul>
     *  <li>OBJECT (IGNORED)</li>
     *  <li>Empty arrays, or arrays with only null values</li>
     * </ul>
     */
```

Method docs should focus on explaining what the function does